### PR TITLE
Merge touch document actions

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
I figured out what this logic was trying to do and i was able to merge the types.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs [hunk 1]
```diff
@@ -15,24 +16,29 @@ internal partial class SolutionCompilationState
 {
     private abstract partial class TranslationAction
     {
-        internal sealed class TouchDocumentAction(
+        internal sealed class TouchDocumentsAction(
             ProjectState oldProjectState,
             ProjectState newProjectState,
-            DocumentState oldState,
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:21]: passing in oldState/oldStates is redundant given that we have oldProjectState.

**Comment by CyrusNajmabadi** [comment]
@ToddGrun ptal.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs [hunk 2]
```diff
@@ -15,24 +16,29 @@ internal partial class SolutionCompilationState
...
-            DocumentState newState)
-            : TranslationAction(oldProjectState, newProjectState)
+            ImmutableArray<DocumentState> newStates) : TranslationAction(oldProjectState, newProjectState)
         {
-            private readonly DocumentState _oldState = oldState;
-            private readonly DocumentState _newState = newState;
+            private readonly ImmutableArray<DocumentState> _oldStates = newStates.SelectAsArray(s => oldProjectState.DocumentStates.GetRequiredState(s.Id));
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:24]: >_oldStates [](http://example.com/codeflow?start=59&length=10)

why have this as a member? Can't each entry just be calculated inside the loop?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs [hunk 3]
```diff
@@ -45,71 +51,18 @@ public override GeneratorDriver TransformGeneratorDriver(GeneratorDriver generat
 
             public override TranslationAction? TryMergeWithPrior(TranslationAction priorAction)
             {
-                if (priorAction is TouchDocumentAction priorTouchAction &&
-                    priorTouchAction._newState == _oldState)
+                if (priorAction is TouchDocumentsAction priorTouchAction &&
+                    priorTouchAction._newStates.SequenceEqual(_oldStates))
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:55]: >priorTouchAction._newStates.SequenceEqual(_oldStates) [](http://example.com/codeflow?start=20&length=53)

Is this more likely to happen than I think it is? (seems like a never would happen thing to me)

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs:55]: it's very likely.  this is what happens when you get a stream of edits coming in for one document (basically, the  typing scenario).

we can either keep making a longer and longer chain of 'edits' where the compilation goes through N steps to get to hte end.  Or we can say that the actions 'merge' and we'll jump from teh starting compilation all the way to teh end step.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski For review when you get back.

**Pull request merged**